src/main/java/database/DataInstancesBuilder.java

```diff
@@ -0,0 +1,119 @@
+        this.batchID = batchID;
+        return this;
+    }
+
+    /**
+    OPTIONAL, if used it just gets data for that subject
+     */
+    public DataInstancesBuilder forSubject(int subjectID) {
+        this.subjectID = subjectID;
+        return this;
+    }
+
+    /**
+     REQUIRED tells which features to put into the dataset
+     Call multiple times to add more features
+     */
+    public DataInstancesBuilder withFeature(int featureID) {
+        features.put(featureID, rndStr.nextString());
+        return this;
+    }
+
+
+
+    public Instances build(Statement stmt) {
+        int numFeatures = features.size();
+        assert numFeatures > 0 : "need atleast one feature";
+        Instances dataset;
+
+        try {
+            String fields = "";
+            String joins = "";
+            String ands = "";
+            String currentAlias;
+            boolean first = true;
+
+            FastVector attributes = new FastVector(numFeatures + 1);
+            final Attribute subjectIdAttribute = new Attribute("subjectID");
+            attributes.addElement(subjectIdAttribute);
+
+            Map<Integer, Attribute> attributeMap = new HashMap<Integer, Attribute>();
+
+            for (Map.Entry<Integer, String> feature : features.entrySet()) {
+                currentAlias = feature.getValue();
+                Attribute a = new Attribute(feature.getKey().toString());
+                attributes.addElement(a);
+                attributeMap.put(feature.getKey(), a);
+
+                if (first) {
+                    first = false;
+                    fields += currentAlias + ".subjectID";
+                    joins += " data as " + currentAlias;
+                } else {
+                    joins += " JOIN data as " + currentAlias + " USING (subjectID, batchID)";
+                }
+
+                fields += ", " + currentAlias + ".featureValue AS " + currentAlias + "fv";
+
+                ands += " AND " + currentAlias + ".featureID = " +  feature.getKey();
+
+            }
+            ResultSet rs = stmt.executeQuery("SELECT " + fields +
+                                             " FROM " + joins +
+                                             " WHERE " + ands.substring(4) + ";");
+
+
+            dataset = new Instances("data", attributes, 10);
+            Instance inst;
+            while (rs.next()) {
+                inst = new Instance(numFeatures + 1);
+                inst.setValue(subjectIdAttribute, 1);
+                for (Map.Entry<Integer, String> feature : features.entrySet()) {
+                    inst.setValue(attributeMap.get(feature.getKey()),
+                                                    rs.getDouble(feature.getValue() + "fv"));
+                }
+
+                dataset.add(inst);
+
+            }
+
+        } catch (SQLException e) {
+            e.printStackTrace();
+            throw new RuntimeException("SQL failed to fetch data");
+        }
+        return dataset;
+    }
```

Comment:
Behemoth function - way too big. Can you refactor into smaller functions with functional names so I know exactly what is going on?